File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.java
class Solution 
{
    int solution(int[][] land) 
    {
        int answer = 0;

        // 각 행의 각 열마다 최대값을 저장해 나간다. 
        int dp[][] = new int[land.length + 1][4];

        // 각 행의 대한 계산
        for(int i = 1; i <= land.length; i++)
        {
            // 각 열에 대한 계산
            for(int j = 0; j < 4; j++)
            {
                // 현재 열을 제외한 나머지 열에 대한 최댓값을 계산한다
                for(int k = 0; k < 4; k++)
                {
                    // 현재 열은 무시
                    if(k == j)
                    {
                        continue;
                    }

                    // 현재 열을 제외한 나머지 열에 대한 계산값을 비교하고 최댓값을 기록한다. 
                    dp[i][j] = Math.max(dp[i][j], land[i - 1][j] + dp[i - 1][k]);

                    answer = Math.max(answer, dp[i][j]);
                }
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.py
def solution(land):
    answer = 0
    n = len(land) # 행
    m = len(land[0]) # 열
    
    for row in range(1, n):
        for col in range(m):
            prev_sum = 0
            
            for prev_col in range(m):
                if prev_col == col:
                    continue
                    
                # 이전행의 다른 열 중 가장 큰 값
                prev_sum = max(prev_sum, land[row-1][prev_col])
                
            land[row][col] += prev_sum

    return max(land[n-1])